dss_datamover/socket_communication.py

```diff
@@ -138,67 +139,62 @@ def recv_json(self, format="JSON", timeout=RECV_TIMEOUT):
-        if msg_len:
+            # first receive message length from payload
+            msg_len_in_bytes = self.socket.recv(MESSAGE_LENGTH)
+            msg_len = int(msg_len_in_bytes)
+
+            # now retrieve rest of payload from buffer based on msg_len
+            # TODO: handle rare edge case where only partial bytes were read before socket termination
             msg_body = b''
-            # Iterate till we receive desired number of bytes.
             received_data_size = 0
             while received_data_size < msg_len:
                 data_size = msg_len - received_data_size
-                try:
-                    received_data = self.socket.recv(data_size)
-                    received_data_size += len(received_data)
-                    msg_body += received_data
-                except socket.timeout as e:
-                    self.logger.error("ClientSocket: Timeout-{}".format(e))
-                    break
-                except socket.error as e:
-                    self.logger.error("ClientSocket: {}".format(e))
+                received_data = self.socket.recv(data_size)
+                received_data_size += len(received_data)
+                msg_body += received_data
 
+            # process received payload
             if msg_body == b'':
                 raise RuntimeError("ClientSocket: Empty message for message length -{}".format(msg_len))
-
-            if msg_body:
+            elif msg_body:
                 if len(msg_body) == msg_len:
                     msg = msg_body.decode("utf8", "ignore")
                 else:
                     self.logger.error("ClientSocket: Received incomplete message.")
-        if format.upper() == "JSON":
-            json_data = {}
-            try:
-                json_data = json.loads(msg)
-            except json.JSONDecodeError as e:
-                self.logger.error("ClientSocket: Bad JSON data - {},{}, {}".format(msg_len, msg, e))
-            except Exception as e:
-                raise Exception("Bad formed message - {}{}, error- {}".format(msg_len, msg, e))
 
-            return json_data
-        else:
-            return msg
+            if format.upper() == "JSON":
+                json_data = {}
+                try:
+                    json_data = json.loads(msg)
+                except json.JSONDecodeError as e:
+                    self.logger.error("ClientSocket: Bad JSON data - {},{}, {}".format(msg_len, msg, e))
+                except Exception as e:
+                    raise Exception("Bad formed message - {}{}, error- {}".format(msg_len, msg, e))
+
+                return json_data
+            else:
+                return msg
+        except socket.timeout as e:
+            self.logger.error("ClientSocket: Timeout ({} seconds) from recv function".format((datetime.now() - time_started).seconds))
+            # return empty status and let the next iteration or worker process the status left on the buffer
+            return json.loads(msg)
```

Comment:
since the **msg** is filled up before and got any exception, you are sending wrong data to the caller. Check again